src/wlapplication_options.cc

```diff
@@ -113,6 +119,254 @@ void set_config_string(const std::string& section,
+                                                           keysym(SDLK_j),
+                                                           "mainmenu_lobby",
+                                                           []() { return _("Metaserver Lobby"); })},
+   {KeyboardShortcut::kMainMenuLogin, KeyboardShortcutInfo(KeyboardShortcutInfo::Scope::kMainMenu,
+                                                           keysym(SDLK_u),
+                                                           "mainmenu_login",
+                                                           []() { return _("Internet Login"); })},
+   {KeyboardShortcut::kMainMenuOptions, KeyboardShortcutInfo(KeyboardShortcutInfo::Scope::kMainMenu,
+                                                             keysym(SDLK_o),
+                                                             "mainmenu_options",
+                                                             []() { return _("Options"); })},
+   {KeyboardShortcut::kGeneralGameBuildhelp,
+    KeyboardShortcutInfo(KeyboardShortcutInfo::Scope::kGameAndEditor,
+                         keysym(SDLK_SPACE),
+                         "buildhelp",
+                         []() { return _("Toggle Buildhelp"); })},
+   {KeyboardShortcut::kGeneralGameMinimap,
+    KeyboardShortcutInfo(KeyboardShortcutInfo::Scope::kGameAndEditor,
+                         keysym(SDLK_m),
+                         "minimap",
+                         []() { return _("Toggle Minimap"); })},
+};
+
+std::string to_string(const KeyboardShortcut id) {
+	return shortcuts_.at(id).descname();
+}
+
+SDL_Keysym get_default_shortcut(const KeyboardShortcut id) {
+	return shortcuts_.at(id).default_shortcut;
+}
+
+static void write_shortcut(const KeyboardShortcut id, const SDL_Keysym code) {
+	set_config_int("keyboard_sym", shortcuts_.at(id).internal_name, code.sym);
+	set_config_int("keyboard_mod", shortcuts_.at(id).internal_name, code.mod);
+}
+bool set_shortcut(const KeyboardShortcut id, const SDL_Keysym code, KeyboardShortcut* conflict) {
+	const KeyboardShortcutInfo::Scope& scope = shortcuts_.at(id).scope;
+	for (auto& pair : shortcuts_) {
+		if (pair.first != id && pair.second.scope == scope && matches_shortcut(pair.first, code)) {
+			*conflict = pair.first;
+			return false;
+		}
+	}
+
+	shortcuts_.at(id).current_shortcut = code;
+	write_shortcut(id, code);
+	return true;
+}
+
+SDL_Keysym get_shortcut(const KeyboardShortcut id) {
+	return shortcuts_.at(id).current_shortcut;
+}
+
+bool matches_shortcut(const KeyboardShortcut id, const SDL_Keysym code) {
+	const SDL_Keysym key = get_shortcut(id);
+	if (key.sym != code.sym) {
+		return false;
+	}
+
+	const bool ctrl1 = code.mod & KMOD_CTRL;
+	const bool shift1 = code.mod & KMOD_SHIFT;
+	const bool alt1 = code.mod & KMOD_ALT;
+	const bool ctrl2 = code.mod & KMOD_CTRL;
+	const bool shift2 = code.mod & KMOD_SHIFT;
+	const bool alt2 = code.mod & KMOD_ALT;
```

Comment:
Shouldn't these use `key.mod`? Simply comparing `code.mod == key.mod` does not work?

Reply:
Thanks, fixed.
'==' comparison is not applicable here because SDL treats e.g. Left Shift and Right Shift as different modifiers.